packages/multichain-account-service/src/MultichainAccountWallet.ts

```diff
@@ -365,122 +374,76 @@ export class MultichainAccountWallet<
    * @returns The discovered accounts for each provider.
    */
   async discoverAndCreateAccounts(): Promise<Record<string, number>> {
-    const providers = this.#providers;
-    const providerContexts = new Map<
-      NamedAccountProvider,
-      ProviderDiscoveryContext
-    >();
-
-    for (const p of providers) {
-      providerContexts.set(p, { stopped: false, groupIndex: 0, count: 0 });
-    }
+    // Start with the next available group index (so we can resume the discovery
+    // from there).
+    let maxGroupIndex = this.getNextGroupIndex();
 
-    let maxGroupIndex = 0;
-
-    const schedule = async (p: NamedAccountProvider, index: number) => {
-      // Ok to cast here because we know that the ctx will always be set.
-      const providerCtx = providerContexts.get(p) as ProviderDiscoveryContext;
-
-      /* istanbul ignore next
-       * Reason: This guard triggers only if `schedule` is invoked for the same
-       * provider while a previous step is still pending. Because the
-       * coordinator self‑reschedules in `finally` and catch‑up scheduling checks
-       * `!qCtx.running`, creating this exact interleaving in unit tests without
-       * racy sleeps is brittle. We keep the guard for safety and exclude it from
-       * coverage.
-       */
-      if (providerCtx.running) {
-        // Wait for current scheduled job before starting a new one.
-        await providerCtx.running.catch((error) => {
-          console.error(error);
-        });
-        providerCtx.running = undefined;
-      }
+    // One serialized loop per provider; all run concurrently
+    const run = async (context: AccountProviderDiscoveryContext) => {
+      const step = (stepName: string) =>
+        `[${context.provider.getName()}] Discover ${stepName} (groupIndex=${context.groupIndex})`;
 
-      providerCtx.groupIndex = index;
+      while (!context.stopped) {
+        // Fast‑forward to current high‑water mark
+        const targetGroupIndex = Math.max(context.groupIndex, maxGroupIndex);
 
-      providerCtx.running = (async () => {
+        console.log(step('STARTED'));
+
+        let accounts: undefined | Bip44Account<KeyringAccount>[];
```

Comment:
Can we do this so we don't need the optional chaining below.
```suggestion
        let accounts: Bip44Account<KeyringAccount>[] = [];
```